subsys/usb/usb_c/usbc_pe_common.c

```diff
@@ -0,0 +1,848 @@
+	const struct device *dev = pe->dev;
+	struct usbc_port_data *data = dev->data;
+	struct protocol_layer_tx_t *prl_tx = data->prl_tx;
+	struct protocol_layer_rx_t *prl_rx = data->prl_rx;
+
+	if (atomic_test_and_clear_bit(&pe->flags, PE_FLAGS_TX_COMPLETE)) {
+		/* Only update data roles if last message sent was Accept */
+		if (prl_tx->msg_type == PD_CTRL_ACCEPT) {
+			/* Update Data Role */
+			pe->data_role = (pe->data_role == TC_ROLE_UFP) ? TC_ROLE_DFP : TC_ROLE_UFP;
+			/* Notify TCPC of role update */
+			tcpc_set_roles(data->tcpc, pe->power_role, pe->data_role);
+			/* Inform Device Policy Manager of Data Role Change */
+			policy_notify(dev, (pe->data_role == TC_ROLE_UFP) ?
+				DATA_ROLE_IS_UFP : DATA_ROLE_IS_DFP);
+		}
+		pe_set_state(dev, PE_SNK_READY);
+	} else if (atomic_test_and_clear_bit(&pe->flags, PE_FLAGS_MSG_DISCARDED)) {
+		/*
+		 * Inform Device Policy Manager that the message was
+		 * discarded
+		 */
+		policy_notify(dev, MSG_DISCARDED);
+		pe_send_soft_reset(dev, prl_rx->emsg.type);
+	}
+}
+
+/**
+ * @brief PE_DRS_Send_Swap Entry state
+ */
+void pe_drs_send_swap_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/* Send Swap DR message */
+	send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_DR_SWAP);
+}
+
+/**
+ * @brief PE_DRS_Send_Swap Run state
+ */
+void pe_drs_send_swap_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+	struct usbc_port_data *data = dev->data;
+	struct protocol_layer_rx_t *prl_rx = data->prl_rx;
+	union pd_header header;
+
+	if (atomic_test_and_clear_bit(&pe->flags, PE_FLAGS_TX_COMPLETE)) {
+		/* Start Sender Response Timer */
+		usbc_timer_start(&pe->pd_t_sender_response);
+	}
+
+	if (atomic_test_and_clear_bit(&pe->flags, PE_FLAGS_MSG_RECEIVED)) {
+		header = prl_rx->emsg.header;
+		if (received_control_message(dev, header, PD_CTRL_REJECT)) {
+			/*
+			 * Inform Device Policy Manager that Data Role Swap
+			 * was Rejected
+			 */
+			policy_notify(dev, MSG_REJECTED_RECEIVED);
+		} else if (received_control_message(dev, header, PD_CTRL_WAIT)) {
+			/*
+			 * Inform Device Policy Manager that Data Role Swap
+			 * needs to Wait
+			 */
+			if (policy_wait_notify(dev, WAIT_DATA_ROLE_SWAP)) {
+				atomic_set_bit(&pe->flags, PE_FLAGS_WAIT_DATA_ROLE_SWAP);
+				usbc_timer_start(&pe->pd_t_wait_to_resend);
+			}
+		} else if (received_control_message(dev, header, PD_CTRL_ACCEPT)) {
+			/* Update Data Role */
+			pe->data_role = (pe->data_role == TC_ROLE_UFP) ? TC_ROLE_DFP : TC_ROLE_UFP;
+			/* Notify TCPC of role update */
+			tcpc_set_roles(data->tcpc, pe->power_role, pe->data_role);
+			/* Inform Device Policy Manager of Data Role Change */
+			policy_notify(dev, (pe->data_role == TC_ROLE_UFP) ?
+				DATA_ROLE_IS_UFP : DATA_ROLE_IS_DFP);
+		} else {
+			/* Protocol Error */
+			policy_notify(dev, PROTOCOL_ERROR);
+			pe_send_soft_reset(dev, PD_PACKET_SOP);
```

Comment:
And now I'm just going to go ahead and add onto this once again...

Reading 8.3.3.4.2.1 PE_SNK_Send_Soft_Reset State while I looked at the soft reset state, apparently we need to perform ErrorRecovery when we hit issues during a DRS.  That one I'd never noticed before, but apparently it's been in the spec since at least 3.0/1.2

Reply:
I've updated the state to perform Error Recovery